File: samples/microsoft/python/enterprise-agent-tutorial/3-production-to-adoption/evaluation_insights.py
"""
Evaluation Insights Analysis Module

Analyzes evaluation results to identify patterns, trends, and improvement opportunities.
"""

import os
import json
from typing import Dict, List
import random
from collections import Counter
import helpers

class EvaluationInsights:
    """Handles evaluation insights and pattern detection."""
    
    def __init__(self, project_client, agent_id: str):
        self.project_client = project_client
        self.agent_id = agent_id
    
    # <insight_generation>
    # NOTE: This code is a non-runnable snippet of the larger sample code from which it is taken.
    def analyze_evaluation_results(self) -> Dict:
        """Generate insights from evaluation data."""
        helpers.print_info("Analyzing evaluation patterns...")
        
        # Simulate evaluation results analysis
        failure_patterns = [
            {
                "pattern": "SharePoint connection timeout",
                "frequency": random.randint(15, 25),
                "impact": "high",
                "recommendation": "Implement retry logic with exponential backoff"
            },
            {
                "pattern": "Incomplete policy citations",
                "frequency": random.randint(8, 15),
                "impact": "medium",
                "recommendation": "Enhance source attribution in responses"
            },
            {
                "pattern": "Outdated Azure documentation",
                "frequency": random.randint(5, 12),
                "impact": "medium",
                "recommendation": "Update MCP server to latest documentation"
            }
        ]
        
        topic_clusters = [
            {
                "cluster": "Azure AD Configuration",
                "question_count": random.randint(120, 180),
                "avg_quality_score": random.uniform(0.75, 0.88),
                "trend": random.choice(["improving", "stable", "declining"])
            },
            {
                "cluster": "Security Policies",
                "question_count": random.randint(80, 130),
                "avg_quality_score": random.uniform(0.70, 0.85),
                "trend": random.choice(["improving", "stable"])
            }
        ]
        
        quality_trends = {
            "current_month": random.uniform(0.80, 0.90),
            "last_month": random.uniform(0.70, 0.80),
            "change_percent": random.uniform(10, 20)
        }
        
        insights = {
            "failure_patterns": failure_patterns,
            "topic_clusters": topic_clusters,
            "quality_trends": quality_trends,
            "recommendations": [
                "Address high-impact failure patterns first",
                "Monitor declining quality trends closely",
                "Expand coverage for emerging topics"
            ]
        }
        
        return insights
    # </insight_generation>
    
    def run(self) -> Dict:
        """Execute evaluation insights analysis."""
        helpers.print_header("Evaluation Insights Analysis")
        
        insights = self.analyze_evaluation_results()
        helpers.print_evaluation_insights(insights)
        helpers.save_json(insights, "evaluation_insights.json")
        
        return insights

def main():
    """Standalone execution."""
    from dotenv import load_dotenv
    load_dotenv()
    
    class MockProjectClient:
        pass
    
    insights = EvaluationInsights(MockProjectClient(), os.getenv("AGENT_ID", "agent-test"))
    insights.run()

if __name__ == "__main__":
    main()

File: samples/microsoft/csharp/enterprise-agent-tutorial/3-production-to-adoption/EvaluationInsights.cs
using System;
using System.Collections.Generic;
using Azure.AI.Projects;

namespace Microsoft.Azure.Samples.ProductionToAdoption
{
    public class EvaluationInsights
    {
        private readonly AIProjectClient? _projectClient;
        private readonly string _agentId;

        public EvaluationInsights(AIProjectClient? projectClient, string agentId)
        {
            _projectClient = projectClient;
            _agentId = agentId;
        }

        // <insight_generation>
        // NOTE: This code is a non-runnable snippet of the larger sample code from which it is taken.
        public Dictionary<string, object> AnalyzeEvaluationResults()
        {
            Helpers.PrintInfo("Analyzing evaluation patterns...");
            var random = new Random();
            var insights = new Dictionary<string, object>
            {
                ["failure_patterns"] = new List<Dictionary<string, object>>
                {
                    new() { ["pattern"] = "SharePoint connection timeout", ["frequency"] = random.Next(15, 25) }
                },
                ["quality_trends"] = new Dictionary<string, object>
                {
                    ["current_month"] = random.NextDouble() * 0.2 + 0.8,
                    ["last_month"] = random.NextDouble() * 0.2 + 0.7
                },
                ["recommendations"] = new List<string> { "Address high-impact failure patterns first" }
            };
            return insights;
        }
        // </insight_generation>

        public Dictionary<string, object> Run()
        {
            Helpers.PrintHeader("Evaluation Insights Analysis");
            var insights = AnalyzeEvaluationResults();
            Helpers.PrintEvaluationInsights(insights);
            Helpers.SaveJson(insights, "evaluation_insights.json");
            return insights;
        }
    }

    public class GatewayIntegration
    {
        private readonly AIProjectClient? _projectClient;
        private readonly string _agentId;

        public GatewayIntegration(AIProjectClient? projectClient, string agentId)
        {
            _projectClient = projectClient;
            _agentId = agentId;
        }

        // <apim_policies>
        // NOTE: This code is a non-runnable snippet of the larger sample code from which it is taken.
        public Dictionary<string, object> ConfigureGatewayPolicies()
        {
            Helpers.PrintInfo("Configuring API Management policies...");
            var config = new Dictionary<string, object>
            {
                ["configuration"] = new Dictionary<string, object>
                {
                    ["endpoint"] = "https://your-apim.azure-api.net/agents/workplace",
                    ["rate_limit"] = "100 requests per minute per user",
                    ["caching_enabled"] = true
                }
            };
            return config;
        }
        // </apim_policies>

        public Dictionary<string, object> Run()
        {
            Helpers.PrintHeader("AI Gateway Integration");
            var config = ConfigureGatewayPolicies();
            Helpers.PrintGatewaySummary(config);
            Helpers.SaveJson(config, "gateway_config.json");
            return config;
        }
    }

    public class ContinuousMonitoring
    {
        private readonly AIProjectClient? _projectClient;
        private readonly string _agentId;

        public ContinuousMonitoring(AIProjectClient? projectClient, string agentId)
        {
            _projectClient = projectClient;
            _agentId = agentId;
        }

        // <quality_monitoring>
        // NOTE: This code is a non-runnable snippet of the larger sample code from which it is taken.
        public Dictionary<string, object> SetupMonitoringSchedule()
        {
            Helpers.PrintInfo("Configuring monitoring schedule...");
            var config = new Dictionary<string, object>
            {
                ["schedule"] = new Dictionary<string, object>
                {
                    ["hourly_checks"] = new Dictionary<string, object> { ["enabled"] = true },
                    ["daily_reports"] = new Dictionary<string, object> { ["enabled"] = true }
                },
                ["alerts"] = new List<Dictionary<string, object>>
                {
                    new() { ["name"] = "Quality Degradation", ["severity"] = "high" }
                }
            };
            return config;
        }
        // </quality_monitoring>

        public Dictionary<string, object> Run()
        {
            Helpers.PrintHeader("Continuous Monitoring Setup");
            var config = SetupMonitoringSchedule();
            Helpers.PrintMonitoringSummary(config);
            Helpers.SaveJson(config, "monitoring_schedule.json");
            return config;
        }
    }

    public class FleetGovernance
    {
        private readonly AIProjectClient? _projectClient;
        private readonly string _agentId;

        public FleetGovernance(AIProjectClient? projectClient, string agentId)
        {
            _projectClient = projectClient;
            _agentId = agentId;
        }

        // <compliance_reporting>
        // NOTE: This code is a non-runnable snippet of the larger sample code from which it is taken.
        public Dictionary<string, object> GenerateGovernanceReport()
        {
            Helpers.PrintInfo("Generating fleet governance report...");
            var random = new Random();
            var report = new Dictionary<string, object>
            {
                ["statistics"] = new Dictionary<string, object>
                {
                    ["total_agents"] = random.Next(12, 18),
                    ["active_users"] = random.Next(120, 180),
                    ["compliance_rate"] = random.NextDouble() * 0.05 + 0.95
                },
                ["recommendations"] = new List<string> { "All agents meet compliance standards" }
            };
            return report;
        }
        // </compliance_reporting>

        public Dictionary<string, object> Run()
        {
            Helpers.PrintHeader("Fleet-Wide Governance");
            var report = GenerateGovernanceReport();
            Helpers.PrintGovernanceSummary(report);
            Helpers.SaveJson(report, "governance_report.json");
            return report;
        }
    }

    public class CostOptimization
    {
        private readonly AIProjectClient? _projectClient;
        private readonly string _agentId;

        public CostOptimization(AIProjectClient? projectClient, string agentId)
        {
            _projectClient = projectClient;
            _agentId = agentId;
        }

        // <cost_analysis>
        // NOTE: This code is a non-runnable snippet of the larger sample code from which it is taken.
        public Dictionary<string, object> AnalyzeCosts()
        {
            Helpers.PrintInfo("Analyzing agent costs...");
            var random = new Random();
            var totalCost = random.NextDouble() * 700 + 2800;
            var report = new Dictionary<string, object>
            {
                ["totals"] = new Dictionary<string, object>
                {
                    ["total_cost"] = totalCost,
                    ["period"] = "Last 30 days"
                },
                ["projected_savings"] = totalCost * 0.3
            };
            return report;
        }
        // </cost_analysis>

        public Dictionary<string, object> Run()
        {
            Helpers.PrintHeader("Cost Monitoring & Optimization");
            var report = AnalyzeCosts();
            Helpers.PrintCostSummary(report);
            Helpers.SaveJson(report, "cost_report.json");
            return report;
        }
    }
}
